### PR TITLE
[SPARK-19311][SQL] fix UDT hierarchy issue 

### PR DESCRIPTION
**Pull request by gmoehler**
## What changes were proposed in this pull request?
acceptType() in UDT will no only accept the same type but also all base types

## How was this patch tested?
Manual test using a set of generated UDTs fixing acceptType() in my user defined types
*Update: I added a test case to https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @rxin Although UDT is now private API, some developers still use it by defining their codes as spark package. I am not sure if you think we need to fix this or not.

**Comment by cloud-fan** [comment]
is it possible to add a unit test? the change LGTM

**Comment by viirya** [comment]
LGTM too. @gmoehler Can you add a unit test?

**Comment by gatorsmile** [comment]
ok to test

**Comment by gatorsmile** [comment]
You can add a test case in https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 1]
```diff
@@ -71,6 +71,94 @@ object UDT {
 
 }
 
+// object and classes to test SPARK-19311
+
+  // Trait/Interface for base type
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Please fix the indentation.

Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: corrected in latest commit ... thanks for pointing out

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 2]
```diff
@@ -71,7 +73,104 @@ object UDT {
...
+  def field: Int
+}
+
+// Trait/Interface for derived type
+@SQLUserDefinedType(udt = classOf[ExampleSubTypeUDT])
+sealed trait IExampleSubType extends IExampleBaseType
+
+// a base class
+class ExampleBaseClass(override val field: Int) extends IExampleBaseType {
+  override def toString: String = field.toString
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: why override `toString` here?

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 3]
```diff
@@ -20,13 +20,15 @@ package org.apache.spark.sql
...
-import org.apache.spark.sql.catalyst.CatalystTypeConverters
+import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow}
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow
 import org.apache.spark.sql.catalyst.util.{ArrayData, GenericArrayData}
 import org.apache.spark.sql.execution.datasources.parquet.ParquetTest
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.test.SharedSQLContext
 import org.apache.spark.sql.types._
 
+
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Nit: Please remove this empty line for avoiding unnecessary changes.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Please simplify it to 
```
class ExampleBaseClass(override val field: Int) extends IExampleBaseType
```

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 4]
```diff
@@ -71,7 +73,104 @@ object UDT {
...
+  override def toString: String = field.toString
+
+}
+
+// a derived class
+class ExampleSubClass(override val field: Int)
+  extends ExampleBaseClass(field) with IExampleSubType
+
+// UDT for base class
+private[spark] class ExampleBaseTypeUDT extends UserDefinedType[IExampleBaseType] {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: `private[spark] ` is not needed here

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 5]
```diff
@@ -71,7 +73,104 @@ object UDT {
...
+
+  override def userClass: Class[IExampleBaseType] = classOf[IExampleBaseType]
+
+  override def hashCode(): Int = classOf[ExampleBaseTypeUDT].getName.hashCode()
+
+  override def equals(other: Any): Boolean = other.isInstanceOf[ExampleBaseTypeUDT]
+
+  override def typeName: String = "exampleBaseType"
+
+  private[spark] override def asNullable: ExampleBaseTypeUDT = this
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Nit: to simplify the test cases, please remove `asNullable `, `typeName `, `equals` and `hashCode`

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 6]
```diff
@@ -71,7 +73,104 @@ object UDT {
...
+// UDT for derived class
+private[spark] class ExampleSubTypeUDT extends UserDefinedType[IExampleSubType] {
+
+  override def sqlType: StructType = {
+    StructType(Seq(
+      StructField("intfield", IntegerType, nullable = false)))
+  }
+
+  override def serialize(obj: IExampleSubType): InternalRow = {
+
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Nit: Please remove this line. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 7]
```diff
@@ -71,7 +73,104 @@ object UDT {
...
+    val row = new GenericInternalRow(1)
+    row.setInt(0, obj.field)
+    row
+  }
+
+  override def deserialize(datum: Any): IExampleSubType = {
+    datum match {
+      case row: InternalRow =>
+        require(row.numFields == 1,
+          s"VectorUDT.deserialize given row with length " +
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Please remove this unneeded `s` String Interpolator. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 8]
```diff
@@ -71,7 +73,104 @@ object UDT {
...
+
+  override def userClass: Class[IExampleSubType] = classOf[IExampleSubType]
+
+  override def hashCode(): Int = classOf[ExampleSubTypeUDT].getName.hashCode()
+
+  override def equals(other: Any): Boolean = other.isInstanceOf[ExampleSubTypeUDT]
+
+  override def typeName: String = "exampleSubType"
+
+  private[spark] override def asNullable: ExampleSubTypeUDT = this
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Nit: to simplify the test cases, please remove asNullable, typeName, equals and hashCode too.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 9]
```diff
@@ -71,7 +73,104 @@ object UDT {
...
+}
+
+
 class UserDefinedTypeSuite extends QueryTest with SharedSQLContext with ParquetTest {
+
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: Nit: Please remove this empty line.

**Comment by gatorsmile** [comment]
It looks great! My comment is just to simplify the unit test cases. Normally, we want to make the unit test cases as simple as possible. Thanks!

LGTM except a few minor comments.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 10]
```diff
@@ -194,4 +293,35 @@ class UserDefinedTypeSuite extends QueryTest with SharedSQLContext with ParquetT
     // call `collect` to make sure this query can pass analysis.
     pointsRDD.as[MyLabeledPoint].map(_.copy(label = 2.0)).collect()
   }
+
+  test("SPARK-19311: UDFs disregard UDT type hierarchy") {
+    UDTRegistration.register(classOf[IExampleBaseType].getName,
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: With `SQLUserDefinedType`, no need to use `UDTRegistration`. We can remove this two lines.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala [hunk 11]
```diff
@@ -71,7 +73,104 @@ object UDT {
...
+        val field = row.getInt(0)
+        new ExampleSubClass(field)
+    }
+  }
+
+  override def userClass: Class[IExampleSubType] = classOf[IExampleSubType]
+
+  override def hashCode(): Int = classOf[ExampleSubTypeUDT].getName.hashCode()
+
+  override def equals(other: Any): Boolean = other.isInstanceOf[ExampleSubTypeUDT]
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: `equals` in `UserDefinedType` has default impl. to call `acceptsType`.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 3 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: ok

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 4 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: ok

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 5 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: ok

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 6 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: not sure which line you mean... we need to overwrite serialize in any case I guess - or is it about a blank line?

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 7 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: ok - replaced it with a fixed string

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 8 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 9 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: ok

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 11 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: right - i have removed equals anyway due to a prev comment of @gatorsmile.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 10 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: i tend to leave them, but remove the @SQLUserDefinedType, so we have a test that uses UDTRegistration

**Comment by gmoehler** [comment]
Thanks for the valuable (and fast!) comments - i have worked them in.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 10 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: oh. if you worry about that, actually we have `UDTRegistrationSuite` for test case of `UDTRegistration`. i am fine to either `SQLUserDefinedType` or `UDTRegistration`.

Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: ok

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 2 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: @gmoehler I think we don't need `toString`?

**Comment by viirya** [comment]
LGTM except one minor comment.

**Comment by gmoehler** [comment]
@viirya Which comment are you referring to? I thought i had included all of them ;-)

**Comment by gatorsmile** [comment]
The remaining comment is: https://github.com/apache/spark/pull/16660#discussion_r97427591

**Comment by gatorsmile** [comment]
LGTM except one comment. Thanks for working on this!

**Comment by viirya** [comment]
@gmoehler `ExampleBaseClass` overrides `toString` impl. That is redundant.

File: sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala (hunk 2 above)
Comment (gmoehler) [inline sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala]: ok

**Comment by gmoehler** [comment]
Thanks for pointing out. I had overseen this comment. 
Really appreciate you commitment to clean code!
Change is pushed and ready for a test build.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.1
